diff/walking/differ.go

```diff
@@ -111,13 +123,14 @@ func (s *walkingDiff) Compare(ctx context.Context, lower, upper []mount.Mount, o
 				}
 			}
 
+			var st *archive.DiffStat
 			if isCompressed {
 				dgstr := digest.SHA256.Digester()
 				compressed, err := compression.CompressStream(cw, compression.Gzip)
 				if err != nil {
 					return errors.Wrap(err, "failed to get compressed stream")
 				}
-				err = archive.WriteDiff(ctx, io.MultiWriter(compressed, dgstr.Hash()), lowerRoot, upperRoot)
+				st, err = archive.WriteDiff(ctx, io.MultiWriter(compressed, dgstr.Hash()), lowerRoot, upperRoot)
```

Comment:
I'm not sure that the channel is good way for this case.

So how about using the `blobStateCalculator` like what the `remote/docker/schema1` does?  It seems it can easy to integrate with `WriteDiff` in the `Compare` function.

```
func (c *blobStateCalculator) Write(p []byte) (int, error) {
        if c.empty {
                for _, b := range p {
                        if b != 0x00 {
                                c.empty = false
                                break
                        }
                }
        }
        return c.digester.Hash().Write(p)
}
```

Reply:
It doesn't seem robust, as theoretically there can be an empty tar with non-zero bytes